wallet-new/src/Cardano/Wallet/Kernel/Addresses.hs

```diff
@@ -0,0 +1,158 @@
+module Cardano.Wallet.Kernel.Addresses (
+    createAddress
+    -- * Errors
+    , CreateAddressError(..)
+    ) where
+
+import           Universum
+
+import           Control.Lens (to)
+import           Formatting (bprint, (%))
+import qualified Formatting as F
+import qualified Formatting.Buildable
+import           System.Random.MWC (GenIO, createSystemRandom, uniformR)
+
+import           Data.Acid (update)
+
+import           Pos.Core (Address, IsBootstrapEraAddr (..), deriveLvl2KeyPair)
+import           Pos.Crypto (EncryptedSecretKey, PassPhrase,
+                     ShouldCheckPassphrase (..))
+
+import           Cardano.Wallet.Kernel (PassiveWallet, walletKeystore, wallets)
+import           Cardano.Wallet.Kernel.DB.AcidState (CreateHdAddress (..))
+import           Cardano.Wallet.Kernel.DB.HdWallet (HdAccountId,
+                     HdAccountIx (..), HdAddressId (..), HdAddressIx (..),
+                     UnknownHdAccount (..), hdAccountIdIx, hdAccountIdParent,
+                     hdAddressIdIx)
+import           Cardano.Wallet.Kernel.DB.HdWallet.Create
+                     (CreateHdAddressError (..), initHdAddress)
+import           Cardano.Wallet.Kernel.DB.HdWallet.Derivation
+                     (HardeningMode (..), deriveIndex)
+import           Cardano.Wallet.Kernel.DB.InDb (InDb (..))
+import qualified Cardano.Wallet.Kernel.Keystore as Keystore
+import           Cardano.Wallet.Kernel.Types (AccountId (..), WalletId (..))
+
+import           Test.QuickCheck (Arbitrary (..), oneof)
+
+data CreateAddressError =
+      CreateAddressUnknownHdAccount UnknownHdAccount
+      -- ^ When trying to create the 'Address', the parent 'Account' was not
+      -- there.
+    | CreateAddressKeystoreNotFound AccountId
+      -- ^ When trying to create the 'Address', the 'Keystore' didn't have
+      -- any secret associated with this 'Account'.
+      -- there.
+    | CreateAddressHdCreationFailed CreateHdAddressError
```

Comment:
We can easily get rid of this one as there are only 2 errors, and one (the duplicate one) we are already handling in the guts of this algorithm, for the second one, we already have a first-class error type to reuse (`CreateAddressUnknownHdAccount`).